_rules/visible-label-in-accessible-name-2ee8b8.md

```diff
@@ -18,77 +18,133 @@ acknowledgements:
     - Anne Thyme Nørregaard
     - Bryn Anderson
     - Jey Nandakumar
+htmlHintIgnore:
+  # https://www.npmjs.com/package/htmlhint
+  # (used with `npm test` to ensure validity of code snippets)
+  - 'title-require'
 ---
 
 ## Applicability
 
-This rule applies to any element that has:
+This rule applies to any HTML or SVG element that:
 
-- a [semantic role](#semantic-role) that is a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget_roles) that [supports name from content](https://www.w3.org/TR/wai-aria-1.1/#namefromcontent), and
-- [visible text content](#visible-text-content), and
-- an `aria-label` or `aria-labelledby` attribute.
+- has a [semantic role](#semantic-role) that is a [widget role](https://www.w3.org/TR/wai-aria-1.1/#widget_roles) that [supports name from content](https://www.w3.org/TR/wai-aria-1.1/#namefromcontent), and
+- has [descendants](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) that are [text nodes](https://www.w3.org/TR/dom/#text) which are both [visible](#visible) as well as [included in the accessibility tree](#included-in-the-accessibility-tree), and
+- has an [accessible name](#accessible-name), even if the name is empty ("").
 
-**Note**: [widget roles](https://www.w3.org/TR/wai-aria-1.1/#widget_roles) that [supports name from content](https://www.w3.org/TR/wai-aria-1.1/#namefromcontent) are: `button`, `checkbox`, `gridcell`, `link`, `menuitem`, `menuitemcheckbox`, `menuitemradio`, `option`, `radio`, `searchbox`, `switch`, `tab`, `treeitem`.
+**Note**: [widget roles](https://www.w3.org/TR/wai-aria-1.1/#widget_roles) that [support name from content](https://www.w3.org/TR/wai-aria-1.1/#namefromcontent) are: `button`, `checkbox`, `gridcell`, `link`, `menuitem`, `menuitemcheckbox`, `menuitemradio`, `option`, `radio`, `switch`, `tab`, `treeitem`.
+
+**Note:** [Text nodes](https://www.w3.org/TR/dom/#text) with only [whitespace](#whitespace) are not [visible](#visible).
 
 ## Expectation
 
-The complete [visible text content](#visible-text-content) of the target element either matches or is contained within its [accessible name][].
+After removing all [punctuation](#punctuation) from both the [visible](#visible) [descendant text content](https://dom.spec.whatwg.org/#concept-descendant-text-content) and the [accessible name](#accessible-name), the [visible](#visible) [descendant text content](https://dom.spec.whatwg.org/#concept-descendant-text-content) of the target element is [included](#included-characters) in its entirety within the [accessible name](#accessible-name) of the element, except for parts of the [visible](#visible) [descendant text content](https://dom.spec.whatwg.org/#concept-descendant-text-content) that do not express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s) and therefore does not live up to the [WCAG definition of text](https://www.w3.org/TR/WCAG21/#dfn-text).
```

Comment:
I've been talking with @kasperisager about this (and my related issues with the definition of visible, see #977).

One way out of it would be to have a definition of "rendered node" (or similar) in the line of:
> a node is rendered if removing it from the DOM tree changes the rendered content.

That way, the `span id="space"` above would contain a rendered text node (with text content `" "`). But the `span id="hidden"` and `span id="off-screen"` would not.

Similarly, an example like https://github.com/act-rules/act-rules.github.io/issues/977#issuecomment-552869788
```html
<p style="background: #white; text-shadow: black 0 0 0px">
  <span style="color: rgba(0,0,0,0); ">Some text in a human language</span>
</p>
```
where the text is technically not visible (only the shadow is) would still have a rendered text node. 

---
The drawbacks with that are:
* Some nodes like above can be invisible but still have rendered text nodes descendants. That's a bit weird but fairly OK (and anyway a consequence of whitespace being invisible).
* If the shadow above is made more blurry, it gets harder and harder to read its content, to the point that it gets hard to argue that the text content should be used (_e.g._ to consider it as a WCAG label).
* Something like `<span style="background: red; color: red;">Foo</span>` has a rendered text node because the `span` stretch for the same width as the text.

(because of these tricky cases and the weird interaction with visible, the definition of rendered can probably use examples…)

All these are pretty edge cases and they can be handled by an assumption in the line of
> This rule assumes that no text node is both rendered and not visible.

(likely needs an explanatory note with examples like spaces)
(essentially, "don't use spaces for layout!")

---
With an assumption like that, this rule can use "visible descendant text nodes" or something similar to grab what is needed.